sherpa/astro/utils/__init__.py

```diff
@@ -375,8 +376,8 @@ def _counts(data, lo, hi, func, *args):
         data.notice()  # clear filter
         data.notice(lo, hi)
         counts = func(*args).sum()
-        data.notice()
     finally:
+        # restore the old filter
```

Comment:
Since you just spend a lot (I guess) of time to understand what exactly is happening, I suggest a few comments, e.g. why do we need to reset the quality filter only it it's not None? If there is no good reason,  suggest to always reset it for simplicity (it uses `getattr(..., ... None)` above so maybe was can do that?

What is this function doing anyway? It starts with an underscore, so we don't need a full doctring, but a 1-line description for te future would be nice.

Reply:
This can be used for a data object that has no quality filter, so we only want to reset it if the field exists and is not None. We have used `getattr` with a default of `None` so this acts as a way to indicate whether we should bother resetting the field.

There are actually two subtle issues here:

- should we do a "dimensionality" check here, so that using a DataIMG object would lead to an error rather than the current behavior of "do something but if you add a filter it is likely to not make sense", but here I just want to ensure we test the current behavior
- should we change the quality array if it is set; I don't think we should but all I want to do is check/test the current behavior and leave the quality-handling to later PRs (as we know that there are issues with out quality handling)

Reply:
For reference, I've created #1444 and #1445 to track these